go/vt/tabletserver/txserializer/tx_serializer.go

```diff
@@ -0,0 +1,220 @@
+package txserializer
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/youtube/vitess/go/vt/vterrors"
+
+	vtrpcpb "github.com/youtube/vitess/go/vt/proto/vtrpc"
+)
+
+// TxSerializer can be used to serialize incoming transactions which target the
+// same row range i.e. table name and WHERE clause are identical.
+// Additional transactions are queued and woken up once it's their turn i.e.
+// they are the first entry in the queue.
+//
+// This implementation is a specialized version of the sync2.Consolidator class.
+// However, there are many substantial differences:
+// - Results are not shared between queued transactions.
+// - Only one waiting transaction and not all are notified when the current one
+//   has finished.
+// - Waiting transactions are woken up in FIFO order.
+// - Waiting transactions are unblocked if their context is done.
+type TxSerializer struct {
+	// Immutable fields.
+	limitGlobal      int
+	limitPerRowRange int
+
+	mu     sync.Mutex
+	queues map[string]*queue
```

Comment:
I think this will be simpler if you used a counting semaphores instead: https://github.com/youtube/vitess/blob/master/go/sync2/semaphore.go.

These should be performed while holding lock on mu:
* If map entry is absent, then create a semaphore with limitPerRowRange items (and obtain the first semaphore).
* If item is already present, try getting a semaphore. If it succeeds, you're allowed to wait.
* If you had the lock and are releasing it, return the semaphore. If the semaphore Size is equal to limitPerRowRange, then no one else is waiting. So, delete the entry.

For the waiting part:
* Create a buffered channel of size 1. The creator goroutine has the lock. So, start the transaction. It's as if you added a "token" to the channel and took it back immediately.
* Once done, return the token to the buffered channel.
* Any goroutine that wants to wait (obtained a semaphore) will wait on this channel. When the lock holder releases the "lock" (return the token to the channel), only one of the waiters will get it. That goroutine then becomes the lock holder.

The waiting part is also effectively a counting semaphore, except that you need to explicitly select on the channel to handle canceled contexts.

I think the above scheme should be a lot simpler. Let me know if it makes sense,